### PR TITLE
STM32Ethernet source files

### PR DESCRIPTION
**Pull request by fprwi6labs**
STM32Ethernet source files for Ethernet library for STM32 cores and Arduino environment.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,22 @@
-# STM32Ethernet
\ No newline at end of file
+# STM32Ethernet
+
+## Ethernet Library for Arduino
+
+With an STM32 board with Ethernet compatibility, this library allows an STM32
```
Comment (fpistm) [inline README.md]: Fix typo: a~~n~~ STM32

File: README.md [hunk 2]
```diff
@@ -1 +1,22 @@
...
+board (NUCLEO, DISCOVERY, ...) to connect to the internet.
+
+For more information about this library please visit us at
+http://www.arduino.cc/en/Reference/Ethernet
+
+## Note
+The library is based on LwIP, a Lightweight TCP/IP stack.
+http://git.savannah.gnu.org/cgit/lwip.git
+
```
Comment (fpistm) [inline README.md]: Mention that LWIP has been ported as Arduino lib and this lib is dependent of this one

File: examples/AdvancedChatServer/AdvancedChatServer.ino [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+ created 18 Dec 2009
+ by David A. Mellis
+ modified 9 Apr 2012
+ by Tom Igoe
+ redesigned to make use of operator== 25 Nov 2013
+ by Norbert Truchsess
+ modified 23 Jun 2017
+ by Wi6Labs
+
```
Comment (fpistm) [inline examples/AdvancedChatServer/AdvancedChatServer.ino]: Add dependency with LWIP lib from stm32duino

File: library.properties [hunk 4]
```diff
@@ -0,0 +1,9 @@
+name=STM32Ethernet
+version=1.0.0
+author=Various
+maintainer=STMicroelectronics
+sentence=Enables network connection (local and Internet) using the STM32 Board.
+paragraph=With this library you can use the STM32 board to connect to Internet. The library provides both Client and server functionalities. The library permits you to connect to a local network also with DHCP and to resolve DNS.
```
Comment (fpistm) [inline library.properties]: Mention LWIP stm23duino dependency

File: src/Dns.cpp [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    stm32_dns_init(iDNSServer.raw_address());
+}
+
+
+int DNSClient::inet_aton(const char* address, IPAddress& result)
+{
+    uint16_t acc = 0; // Accumulator
+    uint8_t dots = 0;
+
+    while (*address)
```
Comment (fpistm) [inline src/Dns.cpp]: check if address is NULL

File: src/Dhcp.h [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+#ifndef Dhcp_h
+#define Dhcp_h
+
+#include "EthernetUdp.h"
+
+/* DHCP state machine. */
+#define STATE_DHCP_STOP				DHCP_OFF
+#define STATE_DHCP_START 			DHCP_START
+#define	STATE_DHCP_DISCOVER		DHCP_WAIT_ADDRESS
```
Comment (fpistm) [inline src/Dhcp.h]: replace tab by space

File: src/Dhcp.h [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+#define DHCP_CHECK_RENEW_FAIL   (1)
+#define DHCP_CHECK_RENEW_OK     (2)
+#define DHCP_CHECK_REBIND_FAIL  (3)
+#define DHCP_CHECK_REBIND_OK    (4)
+
+class DhcpClass {
+private:
+  uint8_t  _dhcpMacAddr[6];
+  unsigned long _timeout;
+	uint8_t _dhcp_lease_state;
```
Comment (fpistm) [inline src/Dhcp.h]: replace tab by space

File: src/Dns.cpp [hunk 8]
```diff
@@ -0,0 +1,105 @@
+// Arduino DNS client for WizNet5100-based Ethernet shield
+// (c) Copyright 2009-2010 MCQN Ltd.
+// Released under Apache License, version 2.0
+
+#include "EthernetUdp.h"
+#include "utility/stm32_eth.h"
+
+#include "Dns.h"
+#include <string.h>
+//#include <stdlib.h>
```
Comment (fpistm) [inline src/Dns.cpp]: remove comment

File: src/Dns.cpp [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    aResult = IPAddress(ipResult);
+
+    return ret;
+}
+
+uint16_t DNSClient::BuildRequest(const char* aName)
+{
+  UNUSED(aName);
+    // Success!  Everything buffered okay
+    return 1;
```
Comment (fpistm) [inline src/Dns.cpp]: please explain why always return 1 (deprecated?)

File: src/Dns.cpp [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+
+uint16_t DNSClient::ProcessResponse(uint16_t aTimeout, IPAddress& aAddress)
+{
+  UNUSED(aTimeout);
+  UNUSED(aAddress);
+    // If we get here then we haven't found an answer
+    return -10;//INVALID_RESPONSE;
```
Comment (fpistm) [inline src/Dns.cpp]: please explain why always return -10 (deprecated?)

File: src/EthernetClient.cpp [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+#include "STM32Ethernet.h"
+#include "EthernetClient.h"
+#include "EthernetServer.h"
+#include "Dns.h"
+
+EthernetClient::EthernetClient() {
+}
+
+EthernetClient::EthernetClient(uint8_t sock) {
+  UNUSED(sock);
```
Comment (fpistm) [inline src/EthernetClient.cpp]: please explain why sock unused (deprecated? compatibilty? not supported?)

File: src/EthernetClient.cpp [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+    return;
+  }
+
+  // Close tcp connection. If failed, force to close connection.
+  if(ERR_OK != tcp_close(_tcp_client->pcb)) {
+    tcp_abort(_tcp_client->pcb);
+  }
+
+  _tcp_client->pcb = NULL;
+  // _tcp_client->state = TCP_NONE;
```
Comment (fpistm) [inline src/EthernetClient.cpp]: to remove

File: src/STM32Ethernet.cpp [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+  /* If there is a local DHCP informs it of our manual IP configuration to
+  prevent IP conflict */
+  stm32_DHCP_manual_config();
+  _dnsServerAddress = dns_server;
+}
+
+int EthernetClass::maintain(){
+  int rc = DHCP_CHECK_NONE;
+
+  // stm32_eth_scheduler();
```
Comment (fpistm) [inline src/STM32Ethernet.cpp]: remove or add more comment

File: src/utility/ethernetif.c [hunk 14]
```diff
@@ -0,0 +1,644 @@
...
+  * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
+  * OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+  * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+  * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
+  * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+  *
+  ******************************************************************************
+  */
+/* Includes ------------------------------------------------------------------*/
+#include "stm32f4xx_hal.h"
```
Comment (fpistm) [inline src/utility/ethernetif.c]: Use generic include stm32_def.h

File: src/utility/stm32_eth.c [hunk 15]
```diff
@@ -0,0 +1,966 @@
...
+      ip_addr_set_zero_ip4(&(gconfig.netmask));
+    #else
+      IP_ADDR4(&(gconfig.netmask),NETMASK_ADDR0,NETMASK_ADDR1,NETMASK_ADDR2,NETMASK_ADDR3);
+    #endif /* LWIP_DHCP */
+  }
+
+  /* Configure the Network interface */
+  Netif_Config();
+
+//TODO: juste temporaire pour debug: laisser à l'utilisateur la possibilité de l'utiliser
```
Comment (fpistm) [inline src/utility/stm32_eth.c]: In English and explain more please

File: src/utility/stm32_eth.c [hunk 16]
```diff
@@ -0,0 +1,966 @@
...
+  else {
+    struct pbuf *q = pbuf_alloc(PBUF_TRANSPORT, size + p->tot_len, PBUF_RAM);
+
+    if(q != NULL) {
+      if(ERR_OK == pbuf_copy(q, p)) {
+        if(ERR_OK == pbuf_take_at(q, (uint8_t *)buffer, size, p->tot_len)) {
+          pbuf_free(p);
+          p = q;
+          return p;
+        } else {
```
Comment (fpistm) [inline src/utility/stm32_eth.c]: empty else ? to remove

**Comment by danieleff** [comment]
`src/utility/stm32_eth.h` includes `#include "stm32f4xx_hal.h"`
Please include `stm32_def.h` instead

**Comment by fprwi6labs** [comment]
Thank you @danieleff .
Fix done.

File: examples/DhcpAddressPrinter/DhcpAddressPrinter.ino [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+  by Wi6Labs
+*/
+
+#include <LwIP.h>
+#include <STM32Ethernet.h>
+
+// Enter a MAC address for your controller below.
+// Newer Ethernet shields have a MAC address printed on a sticker on the shield
+byte mac[] = {
+  0x00, 0xAA, 0xBB, 0xCC, 0xDE, 0x02
```
Comment (fpistm) [inline examples/DhcpAddressPrinter/DhcpAddressPrinter.ino]: Replace by
0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0xED

File: examples/DhcpAddressPrinter/DhcpAddressPrinter.ino [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+  by Arturo Guadalupi
+  modified 23 Jun 2017
+  by Wi6Labs
+*/
+
+#include <LwIP.h>
+#include <STM32Ethernet.h>
+
+// Enter a MAC address for your controller below.
+// Newer Ethernet shields have a MAC address printed on a sticker on the shield
```
Comment (fpistm) [inline examples/DhcpAddressPrinter/DhcpAddressPrinter.ino]: delete this line

File: examples/DhcpChatServer/DhcpChatServer.ino [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+ */
+
+#include <LwIP.h>
+#include <STM32Ethernet.h>
+
+// Enter a MAC address and IP address for your controller below.
+// The IP address will be dependent on your local network.
+// gateway and subnet are optional:
+byte mac[] = {
+  0x00, 0xAA, 0xBB, 0xCC, 0xDE, 0x02
```
Comment (fpistm) [inline examples/DhcpChatServer/DhcpChatServer.ino]: 0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0xED

File: examples/UdpNtpClient/UdpNtpClient.ino [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+ This code is in the public domain.
+
+ */
+
+#include <LwIP.h>
+#include <STM32Ethernet.h>
+#include <EthernetUdp.h>
+
+// Enter a MAC address for your controller below.
+// Newer Ethernet shields have a MAC address printed on a sticker on the shield
```
Comment (fpistm) [inline examples/UdpNtpClient/UdpNtpClient.ino]: delete this line

File: examples/WebClient/WebClient.ino [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+ modified 23 Jun 2017
+ by Wi6Labs
+
+ */
+ 
+#include <LwIP.h>
+#include <STM32Ethernet.h>
+
+// Enter a MAC address for your controller below.
+// Newer Ethernet shields have a MAC address printed on a sticker on the shield
```
Comment (fpistm) [inline examples/WebClient/WebClient.ino]: delete this line

**Pull request merged**